packages/datagrid/src/datagrid.ts

```diff
@@ -2096,6 +2096,77 @@ export class DataGrid extends Widget {
     MessageLoop.postMessage(this._viewport, Private.OverlayPaintRequest);
   }
 
+  private _getMaxWidthInBodyColumn(index: number): number | null {
+    const dataModel = this.dataModel;
+
+    if (!dataModel) {
+      return null;
+    }
+
+    const numRows = dataModel.rowCount('body');
+    const configs = Array.from({ length: numRows }, (_val, idx) =>
+      DataGrid._getConfig(dataModel, idx, index, 'body')
+    );
+
+    // Heuristic: Sort by the length of the text to render and only fully calculate the text width
+    // for the top 100_000 rows by text length.
+    if (numRows > 100_000) {
+      // Sort by descending length.
+      configs.sort(x => -this._getTextToRender(x).length);
+    }
+
+    let maxWidth = 0;
+    for (let i = 0; i < numRows && i < 100_000; ++i) {
+      const textWidth = this._getCellTextWidth(configs[i]);
+      maxWidth = Math.max(maxWidth, textWidth);
+    }
+
+    return maxWidth;
+  }
+
+  private static _getConfig(
+    dataModel: DataModel,
+    row: number,
+    col: number,
+    location: DataModel.CellRegion
+  ): CellRenderer.CellConfig {
+    return {
+      x: 0,
+      y: 0,
+      width: 0,
+      height: 0,
+      region: location,
+      row: row,
+      column: col,
+      value: DataGrid._getCellValue(dataModel, location, row, col),
+      metadata: DataGrid._getCellMetadata(dataModel, location, row, col)
+    };
+  }
+
+  private _getTextToRender(config: CellRenderer.CellConfig): string {
+    const renderer = this.cellRenderers.get(config) as TextRenderer;
+    return renderer.getText(config);
+  }
+
+  private _getCellTextWidth(config: CellRenderer.CellConfig): number {
```

Comment:
This feels a bit like something that should b implemented in each TextRendered instead.